assemblyline_core/dispatching/dispatcher.py

```diff
@@ -623,31 +623,56 @@ def handle_service_results(self):
         queue = self.result_queue
         cpu_mark = time.process_time()
         time_mark = time.time()
+        message_buffer: Dict[str, List] = defaultdict(list)
 
-        while self.running:
-            self.counter.increment_execution_time('cpu_seconds', time.process_time() - cpu_mark)
-            self.counter.increment_execution_time('busy_seconds', time.time() - time_mark)
-
-            message = queue.pop(timeout=1)
+        try:
+            while self.running:
+                self.counter.increment_execution_time('cpu_seconds', time.process_time() - cpu_mark)
+                self.counter.increment_execution_time('busy_seconds', time.time() - time_mark)
 
-            cpu_mark = time.process_time()
-            time_mark = time.time()
+                message = queue.pop(timeout=1)
 
-            if not message:
-                continue
+                cpu_mark = time.process_time()
+                time_mark = time.time()
 
-            sid = message['service_task']['sid']
-            task = self.get_task(sid)
-            if not task:
-                self.log.warning(f'[{sid}] Result returned for finished task.')
-                continue
+                if not message:
+                    continue
 
-            with task.lock:
-                if 'result' in message:
-                    self.process_service_result(task, message['result_key'],
-                                                Result(message['result']), message['temporary_data'])
-                elif 'error' in message:
-                    self.process_service_error(task, message['error_key'], Error(message['error']))
+                # Collect messages binned by sid
+                sid = message['service_task']['sid']
+                message_buffer[sid].append(message)
+
+                # Every time we get a message try each bin, skipping on if we get stuck waiting for the lock
```

Comment:
Isn't it possible that the last message cannot acquire the lock and never get processed?